### PR TITLE
refactor: rename `request.validate` to `request.validateInput`

### PR DESCRIPTION
**Pull request by metcoder95**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [x] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: docs/Reference/Request.md [hunk 1]
```diff
@@ -77,3 +90,161 @@ fastify.post('/:params', options, function (request, reply) {
...
+                        type: 'string' 
+                      } 
+                    } 
+                  })
+validate({ foo: 'bar' }) // true
+
+// or
+
+const validate = request
+                  .getSerializationFunction('body')
```
Comment (Eomm) [inline docs/Reference/Request.md:116]: misprint: the serialization function

File: docs/Reference/Request.md [hunk 2]
```diff
@@ -77,3 +90,161 @@ fastify.post('/:params', options, function (request, reply) {
...
+validate({ foo: 0.5 }) // false
+```
+
+See [.compilaValidationSchema(schema, [httpStatus])](#compileserializationschema)
```
Comment (Eomm) [inline docs/Reference/Request.md:120]:
```suggestion
See [.compilaValidationSchema(schema, [httpStatus])](#compilevalidationschema)
```

File: docs/Reference/Request.md [hunk 3]
```diff
@@ -77,3 +90,161 @@ fastify.post('/:params', options, function (request, reply) {
...
+for more information on how to compile validation function.
+
+### .compileValidationSchema(schema, [httpPart])
+<a id="compilevalidationschema"></a>
+
+This function will compile a validation schema and
+return a function that can be used to validate data.
+The function returned (a.k.a. _validation function_) is compiled
+by using the provided `SchemaControler#ValidationCompiler`.
```
Comment (Eomm) [inline docs/Reference/Request.md:129]: Should it be a link?

File: docs/Reference/Request.md [hunk 4]
```diff
@@ -77,3 +90,161 @@ fastify.post('/:params', options, function (request, reply) {
...
+                        type: 'string' 
+                      } 
+                    } 
+                  })
+validate({ foo: 'bar' }) // true
+
+// or
+
+const validate = request
+                  .compileSerializationSchema({
```
Comment (Eomm) [inline docs/Reference/Request.md:152]: misprint: the serialization is not involved here and it returns a string

File: docs/Reference/Request.md [hunk 5]
```diff
@@ -77,3 +90,161 @@ fastify.post('/:params', options, function (request, reply) {
...
+                        type: 'string' 
+                      } 
+                    } 
+                  }, 200)
+validate({ hello: 'world' }) // false
+```
+
+Note that you should be careful when using this function, as it will cache
+the compiled valiadtion functions based on the schema provided. If the
+schemas provided is mutated or changed, the serialization functions will not
```
Comment (Eomm) [inline docs/Reference/Request.md:165]:
```suggestion
schemas provided are mutated or changed, the validation functions will not
```

File: docs/Reference/Request.md [hunk 6]
```diff
@@ -77,3 +90,161 @@ fastify.post('/:params', options, function (request, reply) {
...
+                      foo: { 
+                        type: 'string' 
+                      } 
+                    } 
+                  }, 200)
+validate({ hello: 'world' }) // false
+```
+
+Note that you should be careful when using this function, as it will cache
+the compiled valiadtion functions based on the schema provided. If the
```
Comment (Eomm) [inline docs/Reference/Request.md:164]:
```suggestion
the compiled validation functions based on the schema provided. If the
```

File: docs/Reference/Request.md [hunk 7]
```diff
@@ -77,3 +90,161 @@ fastify.post('/:params', options, function (request, reply) {
...
+schemas provided is mutated or changed, the serialization functions will not
+detect that the schema has been altered and for instance it will reuse the
+previously compiled serialization function, as the cache is based on
```
Comment (Eomm) [inline docs/Reference/Request.md:167]:
```suggestion
previously compiled validation function, as the cache is based on
```

File: docs/Reference/Request.md [hunk 8]
```diff
@@ -77,3 +90,161 @@ fastify.post('/:params', options, function (request, reply) {
...
+the reference of the schema (Object) previously provided.
+
+If there's a need to change the properties of a schema, always opt to create
+a totally new schema (object), otherwise the implementation won't benefit from
+the cache mechanism.
+
+:Using the following schema as example:
```
Comment (Eomm) [inline docs/Reference/Request.md:174]:
```suggestion
Using the following schema as an example:
```

File: docs/Reference/Request.md [hunk 9]
```diff
@@ -77,3 +90,161 @@ fastify.post('/:params', options, function (request, reply) {
...
+  }
+}
+```
+
+*Not*
+```js 
+const validate = request.compileValidationSchema(schema1)
+
+// Later on...
+schema1.properties.foo.type. = 'integer'
```
Comment (Eomm) [inline docs/Reference/Request.md:191]:
```suggestion
schema1.properties.foo.type = 'integer'
```

File: docs/Reference/Request.md [hunk 10]
```diff
@@ -77,3 +90,161 @@ fastify.post('/:params', options, function (request, reply) {
...
+                    properties: { 
+                      foo: { 
+                        type: 'string' 
+                      } 
+                    } 
+                  }, 200)
+validate({ hello: 'world' }) // false
+```
+
+Note that you should be careful when using this function, as it will cache
```
Comment (Eomm) [inline docs/Reference/Request.md:163]: I wonder if we should just link a blog post (I volunteer to write it) to explain this.

I think it is too technical and the user may loose the focus on the "big picture"

File: docs/Reference/Request.md [hunk 11]
```diff
@@ -77,3 +90,161 @@ fastify.post('/:params', options, function (request, reply) {
...
+      foo: { 
+        type: 'string' 
+      } 
+    } 
+  }, 'body') // true
+
+// or
+
+request
+  .serializeInput({ hello: 'world'}, 'query') // false
```
Comment (Eomm) [inline docs/Reference/Request.md:246]:
```suggestion
  .validateInput({ hello: 'world'}, 'query') // false
```

**Review by Eomm — CHANGES_REQUESTED** [review]

File: docs/Reference/Reply.md [hunk 12]
```diff
@@ -369,7 +369,7 @@ const serialize = reply
 serialize({ foo: 'bar' }) // '{"foo":"bar"}'
 ```
 
-See [.compileSerializationSchema(schema, httpStatus)](#compileserializationschema)
+See [.compileSerializationSchema(schema, [httpStatus])](#compileserializationschema)
```
Comment (mcollina) [inline docs/Reference/Reply.md:372]: why this change?

File: docs/Reference/Reply.md [hunk 13]
```diff
@@ -499,7 +499,7 @@ reply
   .serializeInput({ foo: 'bar'}, 200) // '{"foo":"bar"}'
 ```
 
-See [.compileSerializationSchema(schema, httpStatus)](#compileserializationschema)
+See [.compileSerializationSchema(schema, [httpStatus])](#compileserializationschema)
```
Comment (mcollina) [inline docs/Reference/Reply.md:502]: why this change?

File: docs/Reference/Reply.md (hunk 12 above)
Comment (metcoder95) [reply docs/Reference/Reply.md:372]: Just to indicate that the param is optional. Is that ok or do you think is better indicate it in some other way?

File: docs/Reference/Reply.md (hunk 13 above)
Comment (metcoder95) [reply docs/Reference/Reply.md:502]: As per previous [comment](https://github.com/fastify/fastify/pull/4139#discussion_r922661183)

File: docs/Reference/Request.md (hunk 10 above)
Comment (metcoder95) [reply docs/Reference/Request.md:163]: Agree, didn’t find a way to make it less verbose. Actually I think a Blog post will explain pretty what’s about, happy to be part if I can be helpful somehow 🙂

File: docs/Reference/Request.md (hunk 9 above)
Comment (metcoder95) [reply docs/Reference/Request.md:191]: **Note for myself**: Do not write documentation late in night 😅

File: docs/Reference/Request.md [hunk 14]
```diff
@@ -77,3 +90,161 @@ fastify.post('/:params', options, function (request, reply) {
...
+const validate = request
+                  .compileSerializationSchema({
+                    type: 'object', 
+                    properties: { 
+                      foo: { 
+                        type: 'string' 
+                      } 
+                    } 
+                  }, 200)
+validate({ hello: 'world' }) // false
```
Comment (Eomm) [inline docs/Reference/Request.md:160]: What about renaming it to `validateInput`? (as `serializeInput`)

We did not land this feature yet

**Review by mcollina — APPROVED** [review]
lgtm

File: docs/Reference/Request.md [hunk 15]
```diff
@@ -77,3 +90,161 @@ fastify.post('/:params', options, function (request, reply) {
...
+Note that you should be careful when using this function, as it will cache
+the compiled valiadtion functions based on the schema provided. If the
+schemas provided is mutated or changed, the serialization functions will not
+detect that the schema has been altered and for instance it will reuse the
+previously compiled serialization function, as the cache is based on
+the reference of the schema (Object) previously provided.
+
+If there's a need to change the properties of a schema, always opt to create
+a totally new schema (object), otherwise the implementation won't benefit from
+the cache mechanism.
```
Comment (Fdawgs) [inline docs/Reference/Request.md:172]:
```suggestion
If there is a need to change the properties of a schema, always opt to create
a totally new schema (object), otherwise the implementation will not benefit
from the cache mechanism.
```

Reference docs are formal, so contractions should be avoided. :)

File: docs/Reference/Request.md [hunk 16]
```diff
@@ -77,3 +90,161 @@ fastify.post('/:params', options, function (request, reply) {
...
+
+### .validate(data, [schema | httpStatus], [httpStatus])
+<a id="validate"></a>
+
+This function will validate the input based on the provided schema,
+or http part passed. If both provided, the `httpPart` parameter
+will take presedence.
+
+If there is not a valiodation function for a given `schema`, a new validation
+function will be compiled forwarding the `httpPart` if provided.
```
Comment (Fdawgs) [inline docs/Reference/Request.md:218]:
```suggestion
This function will validate the input based on the provided schema
or HTTP part passed. If both are provided, the `httpPart` parameter
will take precedence.

If there is not a validation function for a given `schema` a new validation
function will be compiled, forwarding the `httpPart` if provided.
```

**Review by Fdawgs — CHANGES_REQUESTED** [review]

File: docs/Reference/Request.md [hunk 17]
```diff
@@ -77,3 +90,161 @@ fastify.post('/:params', options, function (request, reply) {
...
+schemas provided are mutated or changed, the validation functions will not
+detect that the schema has been altered and for instance it will reuse the
+previously compiled validation function, as the cache is based on
+the reference of the schema (Object) previously provided.
+
+If there is a need to change the properties of a schema, always opt to create
+a totally new schema (object), otherwise the implementation will not benefit from
+the cache mechanism.
+
+:Using the following schema as an example:
```
Comment (xtx1130) [inline docs/Reference/Request.md:174]:
```suggestion
Using the following schema as an example:
```

**Review by xtx1130 — CHANGES_REQUESTED** [review]

**Comment by xtx1130** [comment]
Why need to rename `validate` to `validateInput`? This is not mentioned in the pr.

> #### Checklist
> * [x]  run `npm run test` and `npm run benchmark`
> * [ ]  tests and/or benchmarks are included
> * [x]  documentation is changed or added
> * [ ]  commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
>   and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

**Comment by Eomm** [comment]
> Why need to rename `validate` to `validateInput`? This is not mentioned in the pr.

This feature is not yet shipped, so we can change it.

I suggested it because the `reply` object has the new `serializeInput` function to avoid the collision with the `.serialize` one.
So I think it was better to name "equally" the new functions

**Review by Fdawgs — APPROVED** [review]

File: lib/request.js [hunk 18]
```diff
@@ -259,7 +259,7 @@ Object.defineProperties(Request.prototype, {
       return validateFn
     }
   },
-  validate: {
+  validateInput: {
```
Comment (mcollina) [inline lib/request.js:262]: Did validate go out on a release yet?

**Review by mcollina — CHANGES_REQUESTED** [review]

File: lib/request.js (hunk 18 above)
Comment (climba03003) [reply lib/request.js:262]: Not yet.

**Comment by mcollina** [comment]
@climba03003 is this good to land?

**Review by climba03003 — APPROVED** [review]
It is just a refactor on name and document update.

I believe it is good to go.

**Review by xtx1130 — APPROVED** [review]
LGTM
